moshi/src/main/java/com/squareup/moshi/ObjectJsonWriter.java

```diff
@@ -131,7 +140,20 @@ public Object root() {
   }
 
   @Override public JsonWriter value(double value) throws IOException {
-    return value(Double.valueOf(value));
+    long longValue = (long) value;
+    if (Double.compare(longValue, value) == 0) {
+      return value(longValue);
```

Comment:
The reason I care is that `0L` doesn’t equal `0.0d`, and that can be pretty annoying